relay-server/src/utils/envelope_context.rs

```diff
@@ -66,10 +60,44 @@ impl EnvelopeContext {
         self
     }
 
-    /// Records outcomes for all items stored in this context.
+    /// Records an outcome scoped to this envelope's context.
+    ///
+    /// This envelope context should be updated using [`update`](Self::update) soon after this
+    /// operation to ensure that subsequent outcomes are consistent.
+    pub fn track_outcome(&self, outcome: Outcome, category: DataCategory, quantity: usize) {
+        let outcome_aggregator = OutcomeAggregator::from_registry();
+        outcome_aggregator.do_send(TrackOutcome {
+            timestamp: self.received_at,
+            scoping: self.scoping,
+            outcome,
+            event_id: self.event_id,
+            remote_addr: self.remote_addr,
+            category,
+            // Quantities are usually `usize` which lets us go all the way to 64-bit on our
+            // machines, but the protocol and data store can only do 32-bit.
+            quantity: quantity as u32,
+        });
+    }
+
+    /// Accepts the envelope and drops the context.
+    ///
+    /// This should be called if the envelope has been accepted by the upstream, which means that
+    /// the responsibility for logging outcomes has been moved. This function will not log any
+    /// outcomes.
+    pub fn accept(mut self) {
+        if !self.done {
+            self.finish(RelayCounters::EnvelopeAccepted);
+        }
+    }
+
+    /// Records rejection outcomes for all items stored in this context.
     ///
     /// This does not send outcomes for empty envelopes or request-only contexts.
-    pub fn send_outcomes(&self, outcome: Outcome) {
+    pub fn reject(&mut self, outcome: Outcome) {
```

Comment:
what's the reason this takes `&mut self` instead of `mut self`?  I assume this is also the reason you end up with needing the a bit awkward `self.done` flag?  It's a bit unfortunate as it breaks the RAII convention.

Reply:
Unfortunately, there are several places in the code base where we need to pass the envelope context through callbacks such as `map_err()` or hold it on a struct such as `ProcessEnvelopeState`. Right now, having this `&mut` is still more ergonomic and allows to keep most of that code the same for this PR.

There will be a follow-up PR that refactors these places as we'll start to tie envelopes and the context completely together. Hopefully, when we're done with the transition, this can consume `self` again.